### PR TITLE
add more detailed explanations to matrix multiplication programming examples

### PR DESCRIPTION
**Pull request by andrej**
as discussed with Joe

### CODE REVIEW

**Review by jgmelber — APPROVED** [review]

**Comment by jgmelber** [comment]
Thank you! This is great, much more detailed.

**Comment by andrej** [comment]
I forgot to add the note about `ipu_dma_memcpy_nd` stride units. I've added it now

**Comment by jgmelber** [comment]
Can you fix from ipu to npu

File: programming_examples/basic/matrix_multiplication/whole_array/README.md [hunk 1]
```diff
@@ -1,26 +1,228 @@
...
+* After the complete result for the current `m`&times;`n`-block has been calculated, we can release `elem_out`.
+
+### 5. Defining External Data Transfer Sequences
+
+The function signature of the `sequence()` function lists as its arguments all the external buffers from the host that we wish to read from or write to on the AI Engine's shim tiles. The body of this function describes how these buffers are transfered from and to the host, including tiling the input matrices into `m`&times;`k` and `k`&times;`n`-sized sub-matrices, and combining the `m`&times;`n`-sized output tiles into the larger output `M`&times;`N` matrix buffer.
+
+* The `tile_row_block` variable segments the M (rows of A) into smaller chunks, each containing `rows_per_block` tile rows. This is done so the buffer descriptors (BDs) can be reused for efficient DMA transfers.
+* For each column `i`:
+    * For each `tile_row` in the current row block:
+        * The DMA transfer function `ipu_dma_memcpy_nd` loads a segment of matrix A and matrix B data (submatrix a, submatrix b) from the host into the corresponding `inA_fifos` for the respective column, maintaining the appropriate strides and offsets.
```
Comment (jgmelber) [inline programming_examples/basic/matrix_multiplication/whole_array/README.md:208]: `npu_dma_memcpy_nd`

File: programming_examples/basic/matrix_multiplication/whole_array/README.md [hunk 2]
```diff
@@ -1,26 +1,228 @@
...
+        * Analogously to the data layout transformations described [further above](#tiling-and-data-layout-transformations) to translate a `m`&times;`k` matrix into blocks of `r`&times;`s`-submatrices, this transfer translates the input `M`&times;`K` and `K`&times;`N` matrices into submatrices of size `m`&times;`k` and `k`&times;`n`.
+           > Note that data layout transformations in the `ipu_dma_memcpy_nd` operation are expressed in units of 4 bytes. This is why you will see all strides and the lowest-dimension length multiplied by a factor of `word_size_in` or `word_size_out` (to get the size in bytes) and then divided by four (to get the size in units of 4 bytes). This discrepancy will be streamlined in future versions.
+    * The DMA transfer function `ipu_dma_memcpy_nd` sends a segment of matrix C data (submatrix c) from the corresponding `outC_fifos` for the respective column, back to the host while maintaining the appropriate strides and offsets.
```
Comment (jgmelber) [inline programming_examples/basic/matrix_multiplication/whole_array/README.md:211]: `npu_dma_memcpy_nd`

File: programming_examples/basic/matrix_multiplication/whole_array/README.md [hunk 3]
```diff
@@ -1,26 +1,228 @@
...
+* For each column `i`:
+    * For each `tile_row` in the current row block:
+        * The DMA transfer function `ipu_dma_memcpy_nd` loads a segment of matrix A and matrix B data (submatrix a, submatrix b) from the host into the corresponding `inA_fifos` for the respective column, maintaining the appropriate strides and offsets.
+        * Analogously to the data layout transformations described [further above](#tiling-and-data-layout-transformations) to translate a `m`&times;`k` matrix into blocks of `r`&times;`s`-submatrices, this transfer translates the input `M`&times;`K` and `K`&times;`N` matrices into submatrices of size `m`&times;`k` and `k`&times;`n`.
+           > Note that data layout transformations in the `ipu_dma_memcpy_nd` operation are expressed in units of 4 bytes. This is why you will see all strides and the lowest-dimension length multiplied by a factor of `word_size_in` or `word_size_out` (to get the size in bytes) and then divided by four (to get the size in units of 4 bytes). This discrepancy will be streamlined in future versions.
```
Comment (jgmelber) [inline programming_examples/basic/matrix_multiplication/whole_array/README.md:210]: `npu_dma_memcpy_nd`

File: programming_examples/basic/matrix_multiplication/whole_array/README.md [hunk 4]
```diff
@@ -1,26 +1,228 @@
...
+    * The DMA transfer function `ipu_dma_memcpy_nd` sends a segment of matrix C data (submatrix c) from the corresponding `outC_fifos` for the respective column, back to the host while maintaining the appropriate strides and offsets.
+    * After completing DMA transfers for each column, `ipu_sync` is used to synchronize their completion.
```
Comment (jgmelber) [inline programming_examples/basic/matrix_multiplication/whole_array/README.md:212]: `npc_sync`

**Pull request merged**